plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```diff
@@ -0,0 +1,189 @@
+#pragma once
+
+#include <eosio/producer_plugin/snapshot_db_json.hpp>
+
+#include <eosio/chain/block_state.hpp>
+#include <eosio/chain/config.hpp>
+#include <eosio/chain/exceptions.hpp>
+#include <eosio/chain/resource_limits.hpp>
+#include <eosio/chain/resource_limits_private.hpp>
+#include <eosio/chain/transaction.hpp>
+#include <eosio/chain/types.hpp>
+
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/random_access_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+namespace eosio {
+
+namespace bmi = boost::multi_index;
+
+class snapshot_scheduler_listener {
+   virtual void on_start_block(uint32_t height) = 0;
+};
+
+class snapshot_scheduler_handler {
+   virtual void schedule_snapshot(const producer_plugin::snapshot_request_information& sri) = 0;
+   virtual void unschedule_snapshot(uint32_t sri) = 0;
+   virtual producer_plugin::get_snapshot_requests_result get_snapshot_requests() = 0;
+};
+
+class snapshot_scheduler : public snapshot_scheduler_handler, public snapshot_scheduler_listener {
+private:
+   struct by_snapshot_id;
+   struct by_snapshot_value;
+   struct as_vector;
+
+   using snapshot_requests = bmi::multi_index_container<
+         producer_plugin::snapshot_schedule_information,
+         indexed_by<
+               bmi::hashed_unique<tag<by_snapshot_id>, BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_id_information, uint32_t, snapshot_request_id)>,
+               bmi::random_access<tag<as_vector>>,
+               bmi::ordered_unique<tag<by_snapshot_value>,
+                                   composite_key<producer_plugin::snapshot_schedule_information,
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, block_spacing),
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, start_block_num),
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, end_block_num)>>>>;
```

Comment:
I feel that using a `multi-index-container`  is overkill here, and adds unnecessary complexity. I would suggest using an `unordered_set<producer_plugin::snapshot_schedule_information>` with the hash and compare done on the `snapshot_request_id`.

The serialization could as easily be done from the set as from the `<as_vector>` view.
The `<by_snapshot_value>` is used only to find out if a new request is a duplicate, and I think that doing a linear search is probably OK since we are unlikely to have many thousands of snapshot requests.

After all we already do a linear search at every `on_start_block`.

Reply:
@Vlad, I understand that you had a requirement to use `multi_index_container`, and not time to revisit this choice for now, so marking this as resolved.  Thanks!